[expo calendar] add support for complex recurrence rules

tasn opened the pull request:
**NOTE:** This PR is not yet complete. I'm creating it here so @tsapeta can take a look and give feedback (talked to him separately).

# Why

Before this commit we were only supporting basic recurrence rules which
are very limited. This commit adds support for complex recurrence rules.

# How

The JavaScript interface of the basic and complex rules is mutually
compatible. The only difference is that complex rules have more
parameters passed to them.
The relevant iOS docs:
https://developer.apple.com/documentation/eventkit/ekrecurrencerule

# Test Plan

Create events / reminders with the extra parameters, fetch them back afterwards, and verify the created events (both the returned and in the calendar app) look correct.

Example code:
```
const eventId = await Calendar.createEventAsync(containerId, {
  title: 'Test',
  allDay: true,
  startDate: new Date(),
  recurrenceRule: {
    frequency: 'yearly',
    daysOfTheWeek: [{ dayOfTheWeek: 2 }],
    monthsOfTheYears: [2, 3, 5],
  },
});

const event = await Calendar.getEventAsync(eventId);
console.log(event);
```

tasn committed:
[expo calendar] add support for complex recurrence rules

Before this commit we were only supporting basic recurrence rules which
are very limited. This commit adds support for complex recurrence rules.

The JavaScript interface of the basic and complex rules is mutually
compatible. The only difference is that complex rules have more
parameters passed to them.

The relevant iOS docs:
https://developer.apple.com/documentation/eventkit/ekrecurrencerule

tasn commented:
Hey @lukmccall! As I told @tsapeta, I unfortunately don't have access to an iOS build box at the moment (and in the near future), so this patch needs some smoke testing. It's fairly straightforward and just mimics the iOS API. So unless I got the conversions between JS and iOS wrong, it should be good.

Please let me know if there's anything you need from me for testing.

Thanks a lot!

lukmccall reviewed (changes requested):
Hi @tasn, great job 🎊🥇
 
I've tested it. With my requested changes, everything works fine.
However, I have some suggestion for you:
- it will be nice if we export `dayOfTheWeek` and `monthsOfTheYear` as an `enum`. It will be more readable (numbers here are pretty confusing). What do you think? cc @tsapeta.
- please update documentation (pay attention to what happens if you combine new options with old ones) and changelog.md.

Last but not least, what do you want to do with Android?

tasn committed:
RecurrenceRule: fix types and make weekNumber optional

Co-Authored-By: Łukasz Kosmaty <lukasz.kosmaty@student.uj.edu.pl>

tasn commented on packages/expo-calendar/src/Calendar.ts:
```diff
@@ -117,6 +117,13 @@ export type RecurrenceRule = {
+  monthsOfTheYear?: [number];
+  weeksOfTheYear?: [number];
+  daysOfTheYear?: [number];
+  setPositions?: [number];
```
Nice catches, I don't know what's wrong with me, haha. :)

tasn commented:
> Hi @tasn, great job confetti_ball1st_place_medal

Thanks. I'm surprised everything actually works given that it was done blindly. :)

> 
> I've tested it. With my requested changes, everything works fine.

Perfect, will incorporate now.

> However, I have some suggestion for you:
> 
>     * it will be nice if we export `dayOfTheWeek` and `monthsOfTheYear` as an `enum`. It will be more readable (numbers here are pretty confusing). What do you think? cc @tsapeta.

I considered doing it, but decided against it because for all I know the types for the javascript "Date" class also don't define an enum and just use a number, so was going in line with that.

> 
>     * please update documentation (pay attention to what happens if you combine new options with old ones) and changelog.md.

Yes, I will do.

> 
> 
> Last but not least, what do you want to do with Android?

That's up to you. I'd prefer having this in before SDK 37 is out so I can start using it in my iOS app, so if it's possible to get something in with only iOS support, I'd very much like this path.
I already did the research and it's all mappable to Android too so can be added. Android just uses the standard iCal RRULE format in an rrule column.

So, if it's allowed to add iOS only interfaces, that would make my life much easier. If an Android implementation is a requirement, I can take a look into implementing that, but I doubt it will happen in time for SDK 37.

Let me know and I'll act accordingly.

tasn commented:
One more thing: what should I put in the column type for `daysOfTheWeek`? `array` like done elsewhere in the file seems lacking. I'm happy to put `Array<{ dayOfTheWeek: number; weekNumber?: number }>`, but it's not done anywhere else.

lukmccall commented:
> One more thing: what should I put in the column type for `daysOfTheWeek`? `array` like done elsewhere in the file seems lacking. I'm happy to put `Array<{ dayOfTheWeek: number; weekNumber?: number }>`, but it's not done anywhere else.

For dynamic types use `Array<>`. `{ dayOfTheWeek: number; weekNumber?: number }[]` looks weird to me.

lukmccall commented:
I've talked with @tsapeta. We think that `dayOfTheWeek` and `monthsOfTheYear` should be `enums`. 
Please, change it ;)

tasn commented:
> For dynamic types use `Array<>`. `{ dayOfTheWeek: number; weekNumber?: number }[]` looks weird to me.

Sorry, I wasn't clear. I was talking about the docs. What should the type column have in the docs for this field? At the moment I have:
`| daysOfTheWeek | Array<{ dayOfTheWeek: number; weekNumber?: number }> | Optional: the days of the week the event should recur on.  | dayOfTheWeek: 1 to 7 (Sunday to Saturday), weekNumber: -53 to 53 (0 ignores this field, and a negative indicates a value from the end of the range).                               |`

tasn committed:
Recurrence rule: add enums for day/month of the year.

tasn commented:
> I've talked with @tsapeta. We think that `dayOfTheWeek` and `monthsOfTheYear` should be `enums`.
> Please, change it ;)

Done.

lukmccall commented:
> > For dynamic types use `Array<>`. `{ dayOfTheWeek: number; weekNumber?: number }[]` looks weird to me.
> 
> Sorry, I wasn't clear. I was talking about the docs. What should the type column have in the docs for this field? At the moment I have:
> `| daysOfTheWeek | Array<{ dayOfTheWeek: number; weekNumber?: number }> | Optional: the days of the week the event should recur on. | dayOfTheWeek: 1 to 7 (Sunday to Saturday), weekNumber: -53 to 53 (0 ignores this field, and a negative indicates a value from the end of the range). |`

Maybe put in the `type` column `array` and in `description` section adds details about what this array contains.

tasn commented:
> > > For dynamic types use `Array<>`. `{ dayOfTheWeek: number; weekNumber?: number }[]` looks weird to me.
> > 
> > 
> > Sorry, I wasn't clear. I was talking about the docs. What should the type column have in the docs for this field? At the moment I have:
> > `| daysOfTheWeek | Array<{ dayOfTheWeek: number; weekNumber?: number }> | Optional: the days of the week the event should recur on. | dayOfTheWeek: 1 to 7 (Sunday to Saturday), weekNumber: -53 to 53 (0 ignores this field, and a negative indicates a value from the end of the range). |`
> 
> Maybe put `array` in the `type` column and in `description` section adds details about type.

OK, thanks.

tasn committed:
Recurrence rule: update docs for the new complex recurrence rules.

tasn commented:
Just pushed a version that I believe can be merged. I added docs and changed according to all of your comments.

Please let me know if there's anything else you'd like me to change.

Thanks again for your prompt replies.

lukmccall commented on packages/expo-calendar/src/Calendar.ts:
```diff
@@ -112,11 +112,43 @@ export type Alarm = {
+  Thursday,
+  Friday,
+  Saturday,
+}
```
Please add values to all enum fields. In our other modules, we follow this rule. 

Should be:
```
export enum DayOfTheWeek {
  Sunday = 1,
  Monday = 2,
  Tuesday = 3,
  ...
}
```

lukmccall commented on packages/expo-calendar/src/Calendar.ts:
```diff
@@ -112,11 +112,43 @@ export type Alarm = {
+  October,
+  November,
+  December,
+}
```
Same here

lukmccall commented on docs/pages/versions/unversioned/sdk/calendar.md:
```diff
@@ -522,14 +522,26 @@ A person or entity that is associated with an event by being invited or fulfilli
 
 ### RecurrenceRule
 
-A recurrence rule for events or reminders, allowing the same calendar item to recur multiple times.
```
Please add documentation to enum types. It can look like [this](https://docs.expo.io/versions/v35.0.0/sdk/location/#enums) (you can skip the `description` column).

lukmccall commented on docs/pages/versions/unversioned/sdk/calendar.md:
```diff
@@ -522,14 +522,26 @@ A person or entity that is associated with an event by being invited or fulfilli
+[the iCal RFC](https://tools.ietf.org/html/rfc5545#section-3.8.5.3) so you can refer to those to learn more about this
+potentially complex interface.
+
+Not all of the combinations make sense. For example, when frequency is `DAILY`, setting `daysOfTheMonth` makes no sense.
```
Be more precise about the platform. You use here `daysOfTheMonth` which isn't available on Android. Signal to users that Android doesn't have every option yet.

tasn replied:
Each of the added "complex" options is marked as both optional and iOS only, I think it's already descriptive enough, no?

lukmccall replied:
Ok. It's enough :)

lukmccall reviewed (commented):
Great job 👍 
Please, apply requested changes. I think it's the last thing.

lukmccall commented:
Please, don't forget about updating [changelog file](https://github.com/expo/expo/blob/master/CHANGELOG.md).

tasn committed:
Update Changelog.

tasn commented:
Done everything except for the one open question about the docs which I already commented on.

lukmccall commented on docs/pages/versions/unversioned/sdk/calendar.md:
```diff
@@ -552,4 +564,33 @@ A source account that owns a particular calendar. Expo apps will typically not n
+### `Calendar.MonthOfTheYear`
+
+| MonthOfTheYear                      | Value |
+| ----------------------------------- | :---: |
```
I think we don't need to centre this column.
```suggestion
| ----------------------------------- | ----- |
```

lukmccall reviewed (commented):
One more thing :)

tasn committed:
Docs: don't centre MonthOfTheYear's value column

Co-Authored-By: Łukasz Kosmaty <lukasz.kosmaty@student.uj.edu.pl>

tasn commented:
Done.

lukmccall reviewed (approved):
LGTM 💯 🎊 We appreciate your contribution 👍

tasn commented:
:tada: :tada: 

Thanks a lot for all your help!

tsapeta committed:
Merge branch 'master' into rrule

tsapeta reviewed (approved):
Thanks for your contribution! 🚀 💪 
Feel free to send more PRs if you need anything more 😉

tasn commented:
Thanks a lot!

This is not my first patch to Expo and probably won't be my last. ;)
